webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +506,61 @@ case class ProjectADM(
 }
 
 /**
- * The ProjectIdentifierADM factory object, making sure that all necessary checks are performed and all inputs
- * validated and escaped.
+ * Represents the project's identifier, which can be an IRI, shortcode, shortname or UUID.
  */
+sealed trait ProjectIdentifierADM
 object ProjectIdentifierADM {
-  def apply(maybeIri: Option[IRI] = None, maybeShortname: Option[String] = None, maybeShortcode: Option[String] = None)(
-    implicit sf: StringFormatter
-  ): ProjectIdentifierADM = {
-
-    val parametersCount: Int = List(
-      maybeIri,
-      maybeShortname,
-      maybeShortcode
-    ).flatten.size
-
-    // something needs to be set
-    if (parametersCount == 0) throw BadRequestException("Empty project identifier is not allowed.")
-
-    if (parametersCount > 1) throw BadRequestException("Only one option allowed for project identifier.")
-
-    new ProjectIdentifierADM(
-      maybeIri =
-        sf.validateAndEscapeOptionalProjectIri(maybeIri, throw BadRequestException(s"Invalid user project $maybeIri")),
-      maybeShortname = sf.validateAndEscapeOptionalProjectShortname(
-        maybeShortname,
-        throw BadRequestException(s"Invalid user project shortname $maybeShortname")
-      ),
-      maybeShortcode = sf.validateAndEscapeOptionalProjectShortcode(
-        maybeShortcode,
-        throw BadRequestException(s"Invalid user project shortcode $maybeShortcode")
-      )
-    )
+  case class Iri(value: String) extends ProjectIdentifierADM {
+    def apply(value: String): ProjectIri = ProjectIri.make(value).fold(e => throw e.head, v => v)
```

Comment:
tbh, when constructing a `case class` the last thing I would expect that an Exception is thrown.

How about adding the respective four factory methods to the trait which return the Validation and then throw the exception on the callee side of these factory methods, i.e. in the Responders?

Also IMHO the case class should not wrap a `String` but rater the typed identifier, maybe it would make even sense to move the ProjectIdentifier to the shared project?

Reply:
Thanks for the suggestion, I think I messed up here anyway and things are not working as expected... I mean everything is working because validation is not working 🙈